framework/src/modules/validators/api.ts

```diff
@@ -12,20 +12,283 @@
  * Removal or modification of this copyright notice is prohibited.
  */
 
+import { blsPopVerify } from '@liskhq/lisk-cryptography';
 import { BaseAPI } from '../base_api';
-import { ImmutableAPIContext } from '../../node/state_machine';
+import { APIContext, ImmutableAPIContext } from '../../node/state_machine';
+import {
+	EMPTY_KEY,
+	INVALID_BLS_KEY,
+	STORE_PREFIX_BLS_KEYS,
+	STORE_PREFIX_GENERATOR_LIST,
+	STORE_PREFIX_GENESIS_DATA,
+	STORE_PREFIX_VALIDATORS_DATA,
+} from './constants';
+import { APIInitArgs, GeneratorList, GenesisData, ValidatorKeys } from './types';
+import {
+	generatorListSchema,
+	genesisDataSchema,
+	validatorAccountSchema,
+	validatorAddressSchema,
+} from './schemas';
 
 export class ValidatorsAPI extends BaseAPI {
-	// eslint-disable-next-line @typescript-eslint/require-await
-	public async getGenerator(_apiContext: ImmutableAPIContext, _timestamp: number): Promise<Buffer> {
-		return Buffer.alloc(0);
+	private _blockTime!: number;
+
+	public init(args: APIInitArgs) {
+		this._blockTime = args.config.blockTime;
+	}
+
+	public async getGeneratorAtTimestamp(
+		apiContext: ImmutableAPIContext,
+		timestamp: number,
+	): Promise<Buffer> {
+		const genesisData = await this.getGenesisData(apiContext);
+
+		if (timestamp < genesisData.timestamp) {
+			throw new Error('Invalid timestamp');
+		}
+
+		const elapsedTime = timestamp - genesisData.timestamp;
+
+		const generatorListSubStore = apiContext.getStore(this.moduleID, STORE_PREFIX_GENERATOR_LIST);
+		const generatorList = await generatorListSubStore.getWithSchema<GeneratorList>(
+			EMPTY_KEY,
+			generatorListSchema,
+		);
+		const slotIndex = Math.floor(elapsedTime / this._blockTime) % generatorList.addresses.length;
+		return generatorList.addresses[slotIndex];
+	}
+
+	public async getSlotNumber(apiContext: ImmutableAPIContext, timestamp: number): Promise<number> {
+		const genesisData = await this.getGenesisData(apiContext);
+
+		if (timestamp < genesisData.timestamp) {
+			throw new Error('Invalid timestamp');
```

Comment:
```suggestion
			throw new Error('Invalid timestamp.');
```
Maybe we can have more elaborate error message